db/sql/00_msar.sql

```diff
@@ -1013,30 +1061,89 @@ $$ LANGUAGE SQL STABLE RETURNS NULL ON NULL INPUT;
 
 CREATE OR REPLACE FUNCTION
 msar.create_basic_mathesar_user(username text, password_ text) RETURNS TEXT AS $$/*
+Given the username and password_, creates a user on the database server.
+Additionally, grants CREATE, CONNECT and TEMP on the current database to the created user.
+
+Args:
+  username: The name of the user to be created, unquoted.
+  password_: The password for the user to set, unquoted.
 */
-DECLARE
-  sch_name text;
-  mathesar_schemas text[] := ARRAY['mathesar_types', '__msar', 'msar'];
 BEGIN
   PERFORM __msar.exec_ddl('CREATE USER %I WITH PASSWORD %L', username, password_);
   PERFORM __msar.exec_ddl(
     'GRANT CREATE, CONNECT, TEMP ON DATABASE %I TO %I',
     current_database()::text,
     username
   );
+  PERFORM msar.grant_usage_on_mathesar_schemas(username);
+  RETURN username;
+END;
+$$ LANGUAGE plpgsql RETURNS NULL ON NULL INPUT;
+
+
+CREATE OR REPLACE FUNCTION
+msar.grant_usage_on_mathesar_schemas(rolename text) RETURNS void AS $$
+DECLARE
+  sch_name text;
+  mathesar_schemas text[] := ARRAY['mathesar_types', '__msar', 'msar'];
+BEGIN
   FOREACH sch_name IN ARRAY mathesar_schemas LOOP
     BEGIN
-      PERFORM __msar.exec_ddl('GRANT USAGE ON SCHEMA %I TO %I', sch_name, username);
+      PERFORM __msar.exec_ddl('GRANT USAGE ON SCHEMA %I TO %I', sch_name, rolename);
     EXCEPTION
       WHEN invalid_schema_name THEN
         RAISE NOTICE 'Schema % does not exist', sch_name;
     END;
   END LOOP;
-  RETURN username;
 END;
 $$ LANGUAGE plpgsql RETURNS NULL ON NULL INPUT;
 
 
+CREATE OR REPLACE FUNCTION
+msar.create_role(rolename text, password_ text, login_ boolean) RETURNS jsonb AS $$/*
+Creates a login/non-login role, depending on whether the login_ flag is set.
+Only the rolename field is required, the password field is required only if login_ is set to true.
+
+Returns a JSON object describing the created role in the form:
+  {
+    "oid": <int>
+    "name": <str>
+    "super": <bool>
+    "inherits": <bool>
+    "create_role": <bool>
+    "create_db": <bool>
+    "login": <bool>
+    "description": <str|null>
+    "members": <[
+        { "oid": <int>, "admin": <bool> }
+      ]|null>
+  }
+
+Args:
+  rolename: The name of the role to be created, unquoted.
+  password_: The password for the rolename to set, unquoted.
+  login_: Specify whether the role to be created could login.
+*/
+DECLARE
+  sch_name text;
+  mathesar_schemas text[] := ARRAY['mathesar_types', '__msar', 'msar'];
+BEGIN
+  CASE WHEN login_ THEN
+    PERFORM create_basic_mathesar_user(rolename, password_);
+  ELSE
+    PERFORM __msar.exec_ddl('CREATE ROLE %I', rolename);
+    PERFORM __msar.exec_ddl(
+      'GRANT CREATE, TEMP ON DATABASE %I TO %I',
+      current_database()::text,
+      rolename
```

Comment:
The user/role should not have the `CREATE` privilege on the database while creating the role, which both `create_basic_mathesar_user` and the code here grants.

A tangential question:
I was under the impression that we would allow `mathesar_schemas` to be used by `PUBLIC` which will not require us to have to explicitly grant access to each role. Is that not the case? What about roles created outside of Mathesar and configured here? Do those roles not have access to our schemas? (Also a question to @mathemancer)

Reply:
I think this would be the right thing to do. We'd have to add default privileges to them.